interface/requests/wireheader.go

```diff
@@ -6,68 +6,165 @@ package requests
+
+	if err := r.checkForRequest(); err != nil {
+		return err
+	}
+
 	err := binary.Write(buf, binary.LittleEndian, r)
 	return err
 }
 
+func (r *wireHeader) checkForRequest() error {
+	if !isSupportedWireHeaderVersion(r.versionMajor, r.versionMinor) {
+		return fmt.Errorf("invalid version %v.%v", r.versionMajor, r.versionMinor)
+	}
+	if !r.flags.isValid() {
+		return fmt.Errorf("invalid flags %v", r.flags)
+	}
+	if !r.contentType.isValid() {
+		return fmt.Errorf("invalid content type %v", r.contentType)
+	}
+	if !r.acceptType.isValid() {
+		return fmt.Errorf("invalid accept type %v", r.acceptType)
+	}
+	if !r.authType.IsValid() {
+		return fmt.Errorf("invaliid auth type %v", r.authType)
+	}
+	return nil
+}
+
+func isSupportedWireHeaderVersion(maj versionMajorType, min versionMinorType) bool {
+	return maj == versionMajorOne && min == versionMinorZero
+}
+
 const (
 	buffBytes8Bit  int = 1
 	buffBytes16Bit int = 2
 	buffBytes32Bit int = 4
 	buffBytes64Bit int = 8
 )
 
-func (r *WireHeader) parse(buf *bytes.Buffer) error {
+func parseWireHeaderFromBuf(buf *bytes.Buffer) (*wireHeader, error) {
+	r := &wireHeader{}
+	// panic rather than error as this is internal api and this shouldn't happen
+	if buf == nil {
+		panic("buffer pointer is nil")
+	}
 	r.magicNumber = binary.LittleEndian.Uint32(buf.Next(buffBytes32Bit))
 	if r.magicNumber != magicNumber {
-		return errors.New("invalid magic number")
+		return nil, fmt.Errorf("invalid magic number")
 	}
 	r.hdrSize = binary.LittleEndian.Uint16(buf.Next(buffBytes16Bit))
-	if r.hdrSize != responseHeaderSizeValue {
-		logrus.Errorf("Invalid header size (%d != %d)", r.hdrSize, responseHeaderSizeValue)
-		return errors.New("invalid header size")
-	}
-	r.versionMajor = buf.Next(buffBytes8Bit)[0]
-	r.versionMinor = buf.Next(buffBytes8Bit)[0]
-	r.Flags = binary.LittleEndian.Uint16(buf.Next(buffBytes16Bit))
-	r.Provider = buf.Next(buffBytes8Bit)[0]
-	r.Session = binary.LittleEndian.Uint64(buf.Next(buffBytes64Bit))
-	r.ContentType = buf.Next(buffBytes8Bit)[0]
-	r.AcceptType = buf.Next(buffBytes8Bit)[0]
-	r.AuthType = buf.Next(buffBytes8Bit)[0]
-	r.BodyLen = binary.LittleEndian.Uint32(buf.Next(buffBytes32Bit))
-	r.AuthLen = binary.LittleEndian.Uint16(buf.Next(buffBytes16Bit))
-	r.OpCode = binary.LittleEndian.Uint32(buf.Next(buffBytes32Bit))
-	r.Status = binary.LittleEndian.Uint16(buf.Next(buffBytes16Bit))
-	r.Reserved1 = buf.Next(buffBytes8Bit)[0]
-	r.Reserved2 = buf.Next(buffBytes8Bit)[0]
-	return nil
+	if r.hdrSize != wireHeaderSizeValue {
+		return nil, fmt.Errorf("invalid header size (%d != %d)", r.hdrSize, wireHeaderSizeValue)
+	}
+	r.versionMajor = versionMajorType(buf.Next(buffBytes8Bit)[0])
+	r.versionMinor = versionMinorType(buf.Next(buffBytes8Bit)[0])
+	if !isSupportedWireHeaderVersion(r.versionMajor, r.versionMinor) {
+		return nil, fmt.Errorf("unsupported version number %v.%v", r.versionMajor, r.versionMinor)
+	}
+	r.flags = flagsType(binary.LittleEndian.Uint16(buf.Next(buffBytes16Bit)))
```

Comment:
Imagine there were multiple wire protocol versions, then from line 134, you would need to use the correct parsing mechanism of the rest of the header depending on the version. Also the check line 127 might be worth to do here instead, the size of the header might change depending on the version.
I don't want to make your logic too complicated because of this, or too implement everything handling this now, as the wire protocol version might stay the same for a long time, but it's to make the little changes now that would make it easier to handle later

Reply:
fair point, did thing about that at the time - have added to https://github.com/parallaxsecond/parsec-client-go/issues/31